service/internal/checksum/middleware_compute_input_checksum.go

```diff
@@ -264,6 +270,21 @@ func (m *addInputChecksumTrailer) HandleFinalize(
 ) (
 	out middleware.FinalizeOutput, metadata middleware.Metadata, err error,
 ) {
+	var algorithm Algorithm
+	var set bool
+	algorithm, set, err = getInputAlgorithm(ctx)
+	if err != nil {
+		return out, metadata, computeInputTrailingChecksumError{
+			Msg: "failed to get algorithm",
+			Err: err,
+		}
+	}
+
+	checkRequireChecksum(&m.RequireChecksum, m.RequestChecksumCalculation, set)
```

Comment:
I see _why_ you did it this way, and I'm not saying it wouldn't work, but I think we're overcomplicating the way we figure out whether or not we're actually doing the checksum. Basically we've kind of warped the logic here by overloading the meaning of "required" to mean either _literally_ required according to the model, or synethically required if checksum calculation is on. Combined with the addition of the uncomfortable ternary value for requiredness, the ergonomics of this are not where we need them to be.

Basically, all this middleware should be doing is computing the checksum if an algorithm was previously selected. We can determine that in the earlier middleware which actually sets the algorithm on the context - using modeled requiredness and user config to decide accordingly what the algorithm should be.

That's going to give us a much cleaner separation between "should we checksum" and "do checksum" - which as written is fairly tangled.